jetty-jaas/src/main/java/org/eclipse/jetty/jaas/JAASLoginService.java

```diff
@@ -305,53 +277,47 @@ public void logout(UserIdentity user)
      */
     protected String[] getGroups(Subject subject)
     {
+        List<String> roleNameList = Arrays.asList(getRoleClassNames());
+        
         Collection<String> groups = new LinkedHashSet<>();
         Set<Principal> principals = subject.getPrincipals();
         for (Principal principal : principals)
         {
             Class<?> c = principal.getClass();
-            while (c != null)
+            boolean added = false;
+            //check whether the type of this Principle is a role
+            while (c != null && !added)
             {
-                if (roleClassNameMatches(c.getName()))
+                if (roleClassNameMatches(c, roleNameList))
                 {
                     groups.add(principal.getName());
-                    break;
-                }
-
-                boolean added = false;
-                for (Class<?> ci : c.getInterfaces())
-                {
-                    if (roleClassNameMatches(ci.getName()))
-                    {
-                        groups.add(principal.getName());
-                        added = true;
-                        break;
-                    }
-                }
-
-                if (!added)
-                {
-                    c = c.getSuperclass();
+                    added = true;
                 }
                 else
-                    break;
+                    c = c.getSuperclass();
```

Comment:
Maybe this logic of testing the class then moving to the super class should be put inside `roleClassNameMatches`.

Reply:
Done.